packages/affiliates/libs/affiliates/dappmining.js

```diff
@@ -0,0 +1,254 @@
+const { SharedAttributions, Balances } = require("../models");
+
+// Module will calculate the Dapp Mining rewards for a single EMP
+module.exports = ({ queries, empAbi, web3 }) => {
+  assert(queries, "requires queries bigquery interface");
+  assert(empAbi, "requires empAbi");
+  assert(web3, "requires web3");
+
+  const { toBN, fromWei } = web3.utils;
+
+  function getEventStream(empAddress, start, end) {
+    const stream = queries.streamLogsByContract(empAddress, start, end);
+    const decode = DecodeLog(empAbi);
+    return highland(stream).map(log => {
+      return decode(log, {
+        blockNumber: log.block_number,
+        blockTimestamp: moment(log.block_timestamp.value).valueOf(),
+        ...log
+      });
+    });
+  }
+
+  function getAttributionStream(empAddress, start, end) {
+    // We get all attributions from the trace stream. This is very similar to transactions query but
+    // transactions do not contain transactions called from other contracts! Trace does.
+    const stream = queries.streamTracesByContract(empAddress, start, end);
+    const decodeTx = DecodeTransaction(empAbi);
+
+    return (
+      highland(stream)
+        // trace transactions may contain the contract creation event, we cant decode this so filter out
+        // fyi non contract creation "trace_type" are called "call"
+        .filter(tx => tx.trace_type !== "create")
+        .map(tx => {
+          return decodeTx(tx, {
+            blockNumber: tx.block_number,
+            blockTimestamp: moment(tx.block_timestamp.value).valueOf(),
+            ...tx
+          });
+        })
+        // this filters only emp contract "create" calls. We could technically look at more, but these
+        // would be the only "attributed" calls as per the spec
+        .filter(tx => tx.name == "create")
+    );
+  }
+
+  // returns the proportion of rewards given to each developer for a specific point in time
+  function calculateBlockReward(
+    { attributions, balances, whitelist },
+    // return all tags with a percent value for rewards
+    rewards = {}
+  ) {
+    // for every user with a token balance
+    balances.keys().forEach(userid => {
+      // get their stake percent relative to all balances in emp
+      // const balancePercent = balances.getPercent(userid)
+      const balance = balances.get(userid);
+      // console.log({balancePercent,userid})
+      // go through all whitelisted tags
+      whitelist.forEach(tag => {
+        if (rewards[tag] == null) rewards[tag] = "0";
+        // get that tags attribution percent for that user relative to all other affiliates
+        const attributionPercent = attributions.calculateShare(userid, tag);
+        // multiply the balance percent and the attribution percent to get final attribution weight
+        // and add it to the rest of the weights.
+        rewards[tag] = (BigInt(rewards[tag]) + BigInt(balance) * BigInt(attributionPercent)).toString();
```

Comment:
are we sure we want to use BigInt? We explicitly moved away from this in other parts of the affiliates package to conform to the rest of the repo? IMHO we should stick to only using BN.

Reply:
Same comment applies throughout this PR.

Reply:
I will update it to bignum